File: csharp/src/EventBuffer.cs
// The solo event surface (station design 6): a bounded ring buffer plus a
// live tap with serialized callbacks. Events never fail an operation;
// overflow drops oldest and the drop count is visible in Status().
//
// A port of typescript/src/events.ts, which is canonical. Public station
// operations are safe from any thread (design 10.2), so the buffer is
// internally synchronized and tap callbacks run serialized under the
// same lock.

using System;
using System.Collections.Generic;

namespace Voxgig.Station
{
    public class EventBuffer
    {
        private readonly LinkedList<Dictionary<string, object>> ring =
            new LinkedList<Dictionary<string, object>>();
        private readonly int max;
        private long drops = 0;
        private readonly List<Action<Dictionary<string, object>>> taps =
            new List<Action<Dictionary<string, object>>>();
        private readonly object gate = new object();

        public EventBuffer() : this(1000)
        {
        }

        public EventBuffer(int max)
        {
            this.max = max;
        }

        public void Emit(Dictionary<string, object> ev)
        {
            lock (gate)
            {
                ring.AddLast(ev);
                if (ring.Count > max)
                {
                    ring.RemoveFirst();
                    drops++;
                }
                // Serialized, and a throwing tap must not fail the operation
                // that emitted the event.
                foreach (var tap in taps.ToArray())
                {
                    try
                    {
                        tap(ev);
                    }
                    catch (Exception)
                    {
                        // Deliberately swallowed.
                    }
                }
            }
        }

        public List<Dictionary<string, object>> Events()
        {
            lock (gate)
            {
                return new List<Dictionary<string, object>>(ring);
            }
        }

        /// <summary>Subscribe. The returned handle unsubscribes when run.</summary>
        public Action Tap(Action<Dictionary<string, object>> tap)
        {
            lock (gate)
            {
                taps.Add(tap);
            }
            return () =>
            {
                lock (gate)
                {
                    taps.Remove(tap);
                }
            };
        }

        public Dictionary<string, object> Status()
        {
            lock (gate)
            {
                return new Dictionary<string, object>
                {
                    ["buffered"] = ring.Count,
                    ["dropped"] = drops,
                };
            }
        }
    }
}

File: java/src/com/voxgig/station/EventBuffer.java
// The solo event surface (station design 6): a bounded ring buffer plus a
// live tap with serialized callbacks. Events never fail an operation;
// overflow drops oldest and the drop count is visible in status().
//
// A port of typescript/src/events.ts, which is canonical. Public station
// operations are safe from any thread (design 10.2), so the buffer is
// internally synchronized and tap callbacks run serialized under the
// same lock.

package com.voxgig.station;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class EventBuffer {

  private final Deque<Map<String, Object>> ring = new ArrayDeque<>();
  private final int max;
  private long drops = 0;
  private final List<Consumer<Map<String, Object>>> taps = new ArrayList<>();

  public EventBuffer() {
    this(1000);
  }

  public EventBuffer(int max) {
    this.max = max;
  }

  public synchronized void emit(Map<String, Object> event) {
    ring.addLast(event);
    if (ring.size() > max) {
      ring.removeFirst();
      drops++;
    }
    // Serialized, and a throwing tap must not fail the operation that
    // emitted the event.
    for (Consumer<Map<String, Object>> tap : new ArrayList<>(taps)) {
      try {
        tap.accept(event);
      } catch (RuntimeException err) {
        // Deliberately swallowed.
      }
    }
  }

  public synchronized List<Map<String, Object>> events() {
    return new ArrayList<>(ring);
  }

  /** Subscribe. The returned handle unsubscribes when run. */
  public synchronized Runnable tap(Consumer<Map<String, Object>> tap) {
    taps.add(tap);
    return () -> {
      synchronized (this) {
        taps.remove(tap);
      }
    };
  }

  public synchronized Map<String, Object> status() {
    Map<String, Object> out = new LinkedHashMap<>();
    out.put("buffered", ring.size());
    out.put("dropped", drops);
    return out;
  }
}
